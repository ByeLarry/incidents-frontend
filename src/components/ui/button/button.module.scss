@import '../../../index.scss';

.button {
    padding: 0.5rem 1rem;
    border: none;
    background-color: $transparent-color;
    color: $dark-gray;
    cursor: pointer;
    box-sizing: border-box;
    transition: all 0.25s ease-in-out;
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 1px;
    &:hover {
        background-color: $dark-gray;
        color: $light-gray;
        box-shadow: 0 0 0.5rem $dark-gray;
    }

    &:focus {
        outline: none;
        outline-style: solid;
        outline-width: 2px;
        outline-color: $transparent-color;
        box-shadow: 0 0 0.5rem $dark-gray;
        background-color: $dark-gray;
        color: $light-gray;
    }

    &:disabled {
        opacity: 0.5;
        &:hover {
            background-color: $transparent-color;
            color: $dark-gray;
            box-shadow: none;
        }
    }
}