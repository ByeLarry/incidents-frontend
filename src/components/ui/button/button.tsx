import React, { FormEventHandler } from "react";
import styles from "./button.module.scss";

interface ButtonComponentProps {
  className?: string;
  onClick?: () => void;
  children?: React.ReactNode;
  type?: "submit" | "button" | "reset";
  disabled?: boolean;
  title?: string;
  inverse?: boolean;
  textInverse?: boolean;
  noHover?: boolean;
  transparent?: boolean;
  ariaLabel?: string;
  zIndex?: number;
  modalButton?: boolean;
  onSubmit?: FormEventHandler<HTMLButtonElement>;
  form?: string;
}

export const ButtonComponent: React.FC<ButtonComponentProps> = (
  props: ButtonComponentProps
) => {
  return (
    <button
      form={props.form}
      onSubmit={props.onSubmit}
      aria-label={props.ariaLabel}
      type={props.type}
      onClick={props.onClick}
      disabled={props.disabled}
      title={props.title}
      className={`${props.noHover ? styles.button_no_hover : styles.button} ${
        props.modalButton ? styles.button_modal : ""
      } ${props.inverse && styles.button_inverse} ${
        props.textInverse && styles.button__text_inverse
      } ${props.noHover && styles.button_no_hover} ${
        props.transparent && styles.button_transparent
      } ${props.className}`}
      style={{ zIndex: props.zIndex }}
    >
      {props.children}
    </button>
  );
};
