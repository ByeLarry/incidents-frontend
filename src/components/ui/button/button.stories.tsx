import { ButtonComponent } from "./button";
import type { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof ButtonComponent> = {
  component: ButtonComponent,
  title: "UI/Button",
  tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof ButtonComponent>;

export const Modal: Story = {
  args: {
    children: "Button",
    modalButton: true,
    onClick: () => {
      alert("clicked");
    },
  },
};

export const Base: Story = {
    args: {
      children: "Button",
      onClick: () => {
        alert("clicked");
      },
    },
  };
