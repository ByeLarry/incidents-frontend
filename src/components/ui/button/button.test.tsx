import { fireEvent, render, screen } from "@testing-library/react";
import { ButtonComponent } from "./button";
import { vi } from "vitest";

describe("ButtonComponent", () => {
  test("should render button", () => {
    render(<ButtonComponent>Button</ButtonComponent>);
    const button = screen.getByText("Button");
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Button");
    expect(button).toMatchSnapshot();
  });

  test("should render button with props", () => {
    render(
      <ButtonComponent type="button" ariaLabel="Button" modalButton>
        Button
      </ButtonComponent>
    );
    const button = screen.getByText("Button");
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Button");
    expect(button).toMatchSnapshot();
    expect(button).toHaveAttribute("aria-label", "Button");
    expect(button).toHaveAttribute("type", "button");
  });

  test("should render button with onClick and handle click event", () => {
    const handleClick = vi.fn();
    render(
      <ButtonComponent
        type="button"
        ariaLabel="Button"
        modalButton
        onClick={handleClick}
      >
        Button
      </ButtonComponent>
    );
    const button = screen.getByText("Button");
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent("Button");
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(button).toMatchSnapshot();
  });
});
