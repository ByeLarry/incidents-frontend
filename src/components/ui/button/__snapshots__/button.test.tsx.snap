// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ButtonComponent > should render button 1`] = `
<button
  class="_button_522ad7     "
>
  Button
</button>
`;

exports[`ButtonComponent > should render button with onClick and handle click event 1`] = `
<button
  aria-label="Button"
  class="_button_522ad7 _button_modal_522ad7    "
  type="button"
>
  Button
</button>
`;

exports[`ButtonComponent > should render button with props 1`] = `
<button
  aria-label="Button"
  class="_button_522ad7 _button_modal_522ad7    "
  type="button"
>
  Button
</button>
`;
